### PR TITLE
"first measure" in barnumbers is "first in score", not "measure numbered 1"

### PR DESCRIPTION
**Pull request by sammik**
Resolves: https://github.com/musescore/MuseScore/issues/20016

Measure::isFirstInSection() is true, if is first in score, or first after section break

- "Show first" measure numbers depends now on this, and not on "measure numbered one" (first commit)
- when showing all measures numbers and score has anacrusis, "Show first" setting is ignored, and number is showed also on first measure (after anacrusis) (second commit)

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
Doesn't this conflict with #21112?

**Comment by sammik** [comment]
> Doesn't this conflict with #21112?
Yes, it does.

I was thinking about that issue, so as a brainstorming I created my own attempt. 

Was not sure, if it is right to do so, so at least, I didnt mark issue itself.
(May be, it could be found useful once).

File: src/engraving/dom/measure.cpp [hunk 1]
```diff
@@ -533,8 +533,11 @@ bool Measure::showsMeasureNumberInAutoMode()
         return false;
     }
 
+    int interval = style().styleI(Sid::measureNumberInterval);
+
     // Measure numbers should not show on first measure unless specified with Sid::showMeasureNumberOne
-    if (!no()) {
+    // except, when showing numbers on each measure, and first measure is after anacrusis - than show always
```
[the remark this answers is not in the document]
Comment (sammik) [reply src/engraving/dom/measure.cpp]: > than vs. then ;-)

Fresh example of my superior communication skills :D

**Comment by XiaoMigros** [comment]
> I was thinking about that issue, so as a brainstorming I created my own attempt.

Thinking about the issue and creating a PR for it when there is one already open and being worked on isn't exactly the same thing.

**Comment by sammik** [comment]
> Thinking about the issue and creating a PR for it when there is one already open and being worked on isn't exactly the same thing.

@XiaoMigros You are right, please sorry, I appologise. 

Hope, I wouldnt be such stupid anymore.

**Comment by XiaoMigros** [comment]
Well, seems you solved the issue so beats me :)

**Comment by sammik** [comment]
> Well, seems you solved the issue so beats me :)

bitter victory :/

good night

**Review by oktophonie — APPROVED** [review]

File: src/engraving/dom/measure.cpp [hunk 2]
```diff
@@ -533,8 +533,11 @@ bool Measure::showsMeasureNumberInAutoMode()
         return false;
     }
 
+    int interval = style().styleI(Sid::measureNumberInterval);
+
     // Measure numbers should not show on first measure unless specified with Sid::showMeasureNumberOne
-    if (!no()) {
+    // except, when showing numbers on each measure, and first measure is after anacrusis - then show always
+    if (isFirst() || (prevMeasure() && prevMeasure()->irregular() && prevMeasure()->isFirst() && interval != 1)) {
```
Comment (RomanPudashkin) [inline src/engraving/dom/measure.cpp:540]: This doesn't look optimal. We may call prevMeasure() up to 5 times here... You should call it only once

**Comment by sammik** [comment]
@RomanPudashkin rebased (it also fixed vtests)

File: src/engraving/dom/measure.cpp [hunk 3]
```diff
@@ -1930,6 +1933,16 @@ bool Measure::isFirstInSystem() const
     return system()->firstMeasure() == this;
 }
 
+//---------------------------------------------------------
+//   isFirstInSystem
+//---------------------------------------------------------
+
+bool Measure::isFirst() const
```
Comment (cbjeukendrup) [inline src/engraving/dom/measure.cpp:1940]: To prevent confusion, should we rename this method to `isFirstInSection`?

Comment (Jojo-Schmitz) [reply src/engraving/dom/measure.cpp:1940]: And match with the function header (comment)

Comment (sammik) [reply src/engraving/dom/measure.cpp]: Good idea, thanks.

**Review by cbjeukendrup — APPROVED** [review]

**Pull request merged**